### ISSUE TITLE
homie discovery requires three restarts of the MQTT app.

### ISSUE THREAD

**Issue by GledholtHall** [GledholtHall/beta-3#12]
Currently, as of pre24b,  homie discovery must be enabled to  '`Run everytime app starts`' and an app restart then done, afterwards it can be disabled from that should you wish.

Now enable your desired homie discovered devices in the dropdowns and re-run the app again through 'done' - the devices will now appear in HE.  Lastly you will have to run the app yet again so that it creates the MQTT appropriate subscriptions.

Sorry about that as it requires the app be run three times for homie discovered devices to synch up.    I will fix this shortly

**Comment by autolog** [comment]
I think I have followed the above procedure, multiple times but the discovered device isn't appearing in the Devices list. Whne the device was originally discovered, I got these log messages from **MQTT**:

> dev:2162022-01-25 11:06:40.116 debugMQTT> Registering homie device dev-12345678 of type light 
> app:1322022-01-25 11:06:39.991 infoMQTT: Received a homie dim event from dev-12345678 level 0 is indigo-devices
> app:1322022-01-25 11:06:39.618 infoMQTT: Received a homie OnOff event from dev-12345678 turned false
> dev:2162022-01-25 11:05:38.504 debugMQTT> Registering homie device dev-12345678 of type light 
> app:1322022-01-25 11:05:37.979 infoMQTT: Received a homie dim event from dev-12345678 level 0 is indigo-devices
> app:1322022-01-25 11:05:37.865 infoMQTT: Received a homie OnOff event from dev-12345678 turned false

If I look at the MQTT App settings (via gear wheel):
Name: Homie_dim	 Type: enum	Value: ["dev-12345678"]
and under Application State:
Name: dimDevices Value: [dev-12345678]
in case that helps?

My definition in MQTT Explorer is:
![MQTT_Explorer_Definitions_V5](https://user-images.githubusercontent.com/11138074/150969875-86b599b8-be63-490c-bb7d-e150aba727bd.png)

My discovery setup:
![MQTT_Discovery Protocols into HE png](https://user-images.githubusercontent.com/11138074/150970343-e2f4b541-ae2c-4354-a67e-416bdb1228f3.png)

So it looks like the device is discovered but not setup in devices?

Maybe I am missing something in my devices definition in MQTT Explorer?

**Comment by GledholtHall** [comment]
Please can you post a screenshot of the node from MQTT Explorer with all it's sub topics expanded

**Comment by autolog** [comment]
re: Please can you post a screenshot of the node from MQTT Explorer with all it's sub topics expanded

See above. :)

**Comment by GledholtHall** [comment]
You were too quick in responding - I deleted my post when I saw the screenshot but you manage to copy my incriminating post ;-)

**Comment by GledholtHall** [comment]
This is a screenshot from a Hubitat LIFX that does work as I am using it for testing.

![image](https://user-images.githubusercontent.com/63523322/150993673-36975ca8-606f-404d-afb2-441a002422f1.png)

**Comment by autolog** [comment]
I see all your $name are the same (excluding case)  i.e. **What** whereas mine are different. Is that relevant?

**Comment by autolog** [comment]
re: Did homieEnd complete this time (your open issue #11)? = Yes, it dis. Do you want me to close issue #11 ?

**Comment by GledholtHall** [comment]
Yes, please close, names shouldn't matter but I will double check

**Comment by GledholtHall** [comment]
So.. I manually created the device in MQTT Explorer - do make sure all topics are retained. Do not create the set topics.
![image](https://user-images.githubusercontent.com/63523322/151077324-f0db1617-4fad-4bd4-913c-7b2259c415ed.png)
I then enabled 'Run every time app starts' and re-started the app through 'Next/Done'.
![image](https://user-images.githubusercontent.com/63523322/151077187-bbc3d112-2ef4-48ef-9936-ef87ab12c53b.png)
I then ran discovery manually
![image](https://user-images.githubusercontent.com/63523322/151077483-2499a254-cfdb-46c0-be10-453a3322cae8.png)
after it completed I selected the device in the dropdown, and then completed the app run through 'Next/Done'
![image](https://user-images.githubusercontent.com/63523322/151078445-7fa09f4c-35c1-47d7-9775-c094188374a0.png)

Now the new device appeared in the HE devices menu
![image](https://user-images.githubusercontent.com/63523322/151077803-37cd0bef-d467-4227-98ff-a095c387fe69.png)
and the actual device is setup in HE
![image](https://user-images.githubusercontent.com/63523322/151077955-4edabc46-1eb8-4419-b5a1-b21eaeefe90b.png)
at the bottom of that screen in the Data section are the transferred topics
![image](https://user-images.githubusercontent.com/63523322/151078053-940d81fd-b3f2-4ab6-9412-e30edd26a504.png)

This is the log as I discovered the devices from homie.  (You won't see the coloured lines)

![image](https://user-images.githubusercontent.com/63523322/151070997-126477f1-a029-4d68-b3d8-a3936365ad51.png)

**Comment by autolog** [comment]
OK. I will test again tomorrow [Wednesday, 26th]. I will delete the App and Driver and reinstall from scratch just in case something has gone pear shaped. 
I guess I must be doing something wrong?
Will report back during tomorrow.
Thanks for checking it out. 👍

**Comment by GledholtHall** [comment]
Please double check these versions too 
![image](https://user-images.githubusercontent.com/63523322/151079246-6e27b739-3991-4264-a00b-10c3b291051d.png)

**Comment by autolog** [comment]
![Screenshot 2022-01-26 at 00 01 52](https://user-images.githubusercontent.com/11138074/151080127-aaa0a8b6-3906-4355-9d26-0cde17edd161.png)

I have two hubs: home-1 and home-2 , indigo-devices is at the same level. Software versions on the home-1 hub, where I am testing, is the same as in your image.

![Screenshot 2022-01-26 at 00 05 35](https://user-images.githubusercontent.com/11138074/151080438-12af4795-ef45-4153-9a51-21d0e2b1d612.png)

**Comment by GledholtHall** [comment]
Hi Jon,

The best way to check your payloads are being published **retained** is to disconnect and reconnect  MQTT Explorer. The topics/payloads should be all still visible. 

You can double check by selecting the `indigo-devices`  top level topic and then click the wastebin icon in MQTT Explorer which will delete all the topics. Now if you disconnect/reconnect Explorer there will be no indigo-devices device shown and If you then publish your device/nodes from Indigo they will be created again and a disconnect/reconnect from MQTT will once again show them present.

Please note the `…/set` payload should not be published for your devices by yourself… other controllers will do this.   `…/set` must NOT be published retained by any controller.

The key topic for discovery by my app is `$type` as this indicates what type of virtual device I will create.

Currently your device in homie is named indigo-devices - it might be better to name it after the Indigo install eg indigo-1 or whatever so that multiple Indigo instances can be supported. Each device in homie contains many nodes.

Here is the homie protocol spec. just in case you need it..

https://homieiot.github.io/

https://homieiot.github.io/specification/

**Comment by autolog** [comment]
Good morning :)

Thanks for the useful info which I will check out and make sure I am following it.

Am I correct in understanding that for devices published by Indigo for discovery by HE in your MQTT App, that I should just publish the changed value e.g .../onff and payload = true without using /set [as Indigo created the devices]?

**Comment by GledholtHall** [comment]
Yes - those are the state topics.   `…/set` is used for commands TO the device in Indigo.   You will subscribe (listen) to those topics `…/onoff/set` and `…/dim/set` and if anything updates them then you should action that request within Indigo and then update the matching state topic `…/onoff` or …/dim with the updated state payload.

**Comment by GledholtHall** [comment]
As an example and as you have two Hubitat hubs you can publish an internal device from one (complete and compliant homie topics) and discover that device via homie in the other. The virtual device created will stay in synch with the other bidirectionally.

**Comment by autolog** [comment]
Hi Kevin,
First of all, thanks again for all your help, advice and support on this, it is much appreciated. :)

I have discovered that the topics weren't getting retained on my MQTT Mosquitto broker which is running under docker. I have tested with a brew install on my developemnt Mac and they are now. So there will be a slight delay for a day or so for me to get this sorted out and then I will retest everything. 

I thought I should let you know to keep you updated on progress. :)

**Comment by GledholtHall** [comment]
Good - I had that epiphany moment at 6am this morning that non retention would cause this - hence my dawn post...

**Comment by autolog** [comment]
Hi kevin,
Finally got discovery to work on thrre LIFX lamps published from Indigo (as opposed to setting them up in MQTT Explorer). 
I reconfigued Mosquitto outside of Docker and got it configured to ensure retain worked when requested. I also completely deleted the MQTT APP and Driver, rebooted HE, Installed App code and Driver, rebooted and then installed and setup the MQTT App. Following your procedure the three devices appeared in the devices list and so success! 👍  

I will raise a separate issue about device naming.

Now that I have got the basics to work I can continue testing and feedback on developments.
